File: model/src/main/scala/ru/mipt/acsl/decode/model/naming/ElementName.scala
package ru.mipt.acsl.decode.model.naming

/**
  * @author Artem Shein
  */
trait ElementName {
  def asMangledString: String
}

object ElementName {

  private case class Impl(value: String) extends ElementName {

    override def asMangledString: String = value

  }

  def mangleName(name: String): String = {
    var result = name
    if (result.startsWith("^")) {
      result = result.substring(1)
    }
    result = "[ \\\\^]".r.replaceAllIn(result, "")
    if (result.isEmpty)
      sys.error("invalid name")
    result
  }

  def newFromSourceName(name: String): ElementName = Impl(ElementName.mangleName(name))

  def newFromMangledName(name: String): ElementName = Impl(name)
}

File: model/src/main/java/ru/mipt/acsl/decode/model/naming/ElementName.java
package ru.mipt.acsl.decode.model.naming;

/**
 * Created by metadeus on 06.06.16.
 */
public interface ElementName {

    static String mangleName(String sourceName) {
        String result = sourceName;
        if (result.startsWith("^")) {
            result = result.substring(1);
        }
        result = result.replaceAll("[ \\\\\\^]", "");
        if (result.isEmpty()) {
            throw new IllegalArgumentException("invalid name");
        }
        return result;
    }

    static ElementName newInstanceFromMangledName(String mangledName) {
        return new ElementNameImpl(mangledName);
    }

    static ElementName newInstanceFromSourceName(String sourceName) {
        return new ElementNameImpl(mangleName(sourceName));
    }

    String mangledNameString();

}
